consensus/src/block_graph.rs

```diff
@@ -3727,6 +3727,22 @@ impl BlockGraph {
             }
         }
 
+        // retain extra history according to the config
+        // this is useful to avoid desync on temporary connection loss
+        for a_block in self.active_index.iter() {
```

Comment:
is this ok to go through every active block ?

Reply:
there is no index by slot, and we already do it a few times elsewhere in the prune_active method. So it doesn't change the overall complexity of the method, it just adds one to the complexity pre-factor

Reply:
We do it only once before, why don't you put everything you need in that one loop ?

Reply:
The force-keep part needs to be done after "grow with parents & fill thread holes twice", otherwise we will also "grow" the force-kept set and try to fill holes that cannot exist there

Reply:
If you have something special in mind, feel free to propose it

Reply:
You could fill `to retain` with this criterion and a `to_grow` variable that would only have the blocks we want to grow dependencies from. And once `to_grow` has grow you can extend `to_retain` with it.

Reply:
would that improve performance and readability ? looks like one more memory allocation

Reply:
But one less loop on everything